File: bench/astar/astar.hs
{-# LANGUAGE BangPatterns #-}
-- A* on a 100×100 grid — mutable IOUArray implementation.
-- Uses the same wall rule and sorted-list open set as Sprout.
-- g_score and closed use IOUArray Int for O(1) indexed updates.
-- Timing: wall-clock via Data.Time.Clock.System (monotonic, nanosecond precision).
--
-- Compile: ghc -O2 -o astar_hs astar.hs
-- Run:     ./astar_hs
import Data.Array.IO
import Data.Time.Clock.System (getSystemTime, SystemTime(..))
import Text.Printf

w, goalX, goalY, inf, iters :: Int
w = 100; goalX = 99; goalY = 99; inf = 9999999; iters = 60000

isWall :: Int -> Int -> Bool
isWall x y
  | x <= 0 || y <= 0 || x >= goalX || y >= goalY = False
  | otherwise = (x * 5 + y * 3) `mod` 13 < 4

heur :: Int -> Int -> Int
heur x y = abs (goalX - x) + abs (goalY - y)

-- Sorted insertion ascending by f.
openInsert :: (Int,Int,Int,Int) -> [(Int,Int,Int,Int)] -> [(Int,Int,Int,Int)]
openInsert e@(!f,_,_,_) [] = [e]
openInsert e@(!f,_,_,_) all@(h@(!hf,_,_,_):rest)
  | f <= hf   = e : all
  | otherwise = h : openInsert e rest

astar :: IO Int
astar = do
    gScore <- newArray (0, w*w-1) inf :: IO (IOUArray Int Int)
    closed <- newArray (0, w*w-1) False :: IO (IOUArray Int Bool)
    writeArray gScore 0 0
    go [(heur 0 0, 0, 0, 0)] gScore closed
  where
    go [] _ _ = return (-1)
    go ((_,!g,!x,!y):rest) gs cl = do
      if x == goalX && y == goalY
        then return g
        else do
          let idx = y*w+x
          isCl <- readArray cl idx
          if isCl
            then go rest gs cl
            else do
              writeArray cl idx True
              open' <- foldl (stepNeighbor g gs cl) (return rest)
                         [(x,y-1),(x,y+1),(x-1,y),(x+1,y)]
              go open' gs cl

    stepNeighbor !g gs cl macc (!nx,!ny)
      | nx < 0 || nx >= w || ny < 0 || ny >= w = macc
      | isWall nx ny = macc
      | otherwise = do
          acc <- macc
          let nidx = ny*w+nx
          isCl <- readArray cl nidx
          if isCl then return acc
          else do
            eg <- readArray gs nidx
            let !ng = g + 1
            if ng < eg
              then do
                writeArray gs nidx ng
                return $! openInsert (ng + heur nx ny, ng, nx, ny) acc
              else return acc

toNs :: SystemTime -> Integer
toNs (MkSystemTime s ns) = fromIntegral s * 1_000_000_000 + fromIntegral ns

benchmark :: IO ()
benchmark = do
    -- warmup
    mapM_ (const astar) [1..5 :: Int]
    t0 <- getSystemTime
    total <- sum <$> mapM (const astar) [1..iters :: Int]
    t1 <- getSystemTime
    let ms = fromIntegral (toNs t1 - toNs t0) / 1_000_000 :: Double
        usPerRun = round (ms * 1000 / fromIntegral iters) :: Int
    printf "A* 100x100, %d runs: %.1f ms  (%d us/run, path=%d steps)\n"
           iters ms usPerRun (total `div` iters)

main :: IO ()
main = benchmark

File: bench/astar/astar.go
// A* on a 100×100 grid — sorted-slice open set, mutable int arrays.
// Same wall rule and sorted-list open set as the Sprout implementation.
// g_score and closed use flat [10000]int / [10000]bool for O(1) access.
// Open-set insert: extend by 1 then in-place copy-shift — O(n) time, O(1) amortised alloc.
//
// Run: go run astar.go
package main

import (
	"fmt"
	"time"
)

const (
	W     = 100
	GoalX = 99
	GoalY = 99
	Inf   = 9999999
	Iters = 600
)

func isWall(x, y int) bool {
	if x <= 0 || y <= 0 || x >= GoalX || y >= GoalY {
		return false
	}
	return (x*5+y*3)%13 < 4
}

func heur(x, y int) int {
	dx := GoalX - x
	if dx < 0 {
		dx = -dx
	}
	dy := GoalY - y
	if dy < 0 {
		dy = -dy
	}
	return dx + dy
}

type entry struct{ f, g, x, y int }

// Sorted slice — O(n) insert, same asymptotic class as Sprout's sorted list.
type openSet []entry

func (o *openSet) insert(e entry) {
	*o = append(*o, entry{}) // extend by 1 (amortised; reuses backing array when cap > len)
	s := *o
	i := 0
	for i < len(s)-1 && s[i].f <= e.f {
		i++
	}
	copy(s[i+1:], s[i:len(s)-1]) // shift tail right in place
	s[i] = e
}

var dirs = [4][2]int{{0, -1}, {0, 1}, {-1, 0}, {1, 0}}

func astar() int {
	var gScore [W * W]int
	var closed [W * W]bool
	for i := range gScore {
		gScore[i] = Inf
	}
	gScore[0] = 0

	open := openSet{{heur(0, 0), 0, 0, 0}}

	for len(open) > 0 {
		cur := open[0]
		open = open[1:]
		x, y, g := cur.x, cur.y, cur.g

		if x == GoalX && y == GoalY {
			return g
		}
		idx := y*W + x
		if closed[idx] {
			continue
		}
		closed[idx] = true

		for _, d := range dirs {
			nx, ny := x+d[0], y+d[1]
			if nx < 0 || nx >= W || ny < 0 || ny >= W {
				continue
			}
			nidx := ny*W + nx
			if isWall(nx, ny) || closed[nidx] {
				continue
			}
			ng := g + 1
			if ng < gScore[nidx] {
				gScore[nidx] = ng
				open.insert(entry{ng + heur(nx, ny), ng, nx, ny})
			}
		}
	}
	return -1
}

func main() {
	// warmup
	for i := 0; i < 5; i++ {
		astar()
	}

	t := time.Now()
	total := 0
	for i := 0; i < Iters; i++ {
		g := astar()
		if g >= 0 {
			total += g
		}
	}
	elapsed := time.Since(t)
	ms := float64(elapsed.Microseconds()) / 1000.0
	usPerRun := elapsed.Microseconds() / int64(Iters)
	fmt.Printf("A* 100x100, %d runs: %.1f ms  (%d us/run, path=%d steps)\n",
		Iters, ms, usPerRun, total/Iters)
}
